src/Shared/Recruit.Vacancies.Client/Application/CommandHandlers/CloneVacancyCommandHandler.cs

```diff
@@ -0,0 +1,90 @@
+using MediatR;
+using System.Threading;
+using System.Threading.Tasks;
+using Esfa.Recruit.Vacancies.Client.Application.Services;
+using Esfa.Recruit.Vacancies.Client.Domain.Entities;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using Esfa.Recruit.Vacancies.Client.Domain.Exceptions;
+
+namespace Esfa.Recruit.Vacancies.Client.Application.CommandHandlers
+{
+    public class CloneVacancyCommandHandler: IRequestHandler<CloneVacancyCommand>
+    {
+        private readonly ILogger<CloneVacancyCommandHandler> _logger;
+        private readonly IVacancyRepository _repository;
+        private readonly IMessaging _messaging;
+        private readonly ITimeProvider _timeProvider;
+
+        public CloneVacancyCommandHandler(
+            ILogger<CloneVacancyCommandHandler> logger,
+            IVacancyRepository repository, 
+            IMessaging messaging, 
+            ITimeProvider timeProvider)
+        {
+            _logger = logger;
+            _repository = repository;
+            _messaging = messaging;
+            _timeProvider = timeProvider;
+        }
+
+        public async Task Handle(CloneVacancyCommand message, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Cloning new vacancy with id: {vacancyId} from vacancy with id: {clonedVacancyId}", message.IdOfVacancyToClone, message.NewVacancyId);
+
+            var vacancy = await _repository.GetVacancyAsync(message.IdOfVacancyToClone);
+
+            if (vacancy.Status != VacancyStatus.Submitted && vacancy.Status != VacancyStatus.Live && vacancy.Status != VacancyStatus.Closed)
+            {
+                _logger.LogError($"Unable to clone vacancy {{vacancyId}} due to it having a status of {vacancy.Status}.", message.IdOfVacancyToClone);
+                
+                throw new InvalidStateException($"Vacancy is not in correct state to be cloned. Current State: {vacancy.Status}");
+            }
+
+            var clone = CreateClone(message, vacancy);
+
+            await _repository.CreateAsync(clone);
+
+            await _messaging.PublishEvent(new VacancyCreatedEvent
+            {
+                EmployerAccountId = clone.EmployerAccountId,
+                VacancyId = clone.Id
+            });
+        }
+
+        private Vacancy CreateClone(CloneVacancyCommand message, Vacancy vacancy)
+        {
+            var now = _timeProvider.Now;
+
+            var clone = JsonConvert.DeserializeObject<Vacancy>(JsonConvert.SerializeObject(vacancy));
+
+            // Properties to replace
+            clone.Id = message.NewVacancyId;
+            clone.CreatedByUser = message.User;
+            clone.CreatedDate = now;
+            clone.LastUpdatedByUser = message.User;
+            clone.LastUpdatedDate = now;
+            clone.SourceOrigin = SourceOrigin.EmployerWeb;
+            clone.SourceType = SourceType.Clone;
+            clone.SourceVacancyReference = vacancy.VacancyReference;
```

Comment:
Should we trap the source Vacancy Id guid?

Reply:
Do we need to if we've got the VacancyReference? The conversation with @mgwilliam around regeneration of documents relied on the fact that we don't have store internal identifiers.

Reply:
Ignore me. For some reason I thought we would lose track of which 'version' of a vacancy was used for the clone but of cause the clone gets a unique VacancyReference.